### PR TITLE
Use appender.Remote in place of appender.Appender

### PR DESCRIPTION
**Pull request by gdbelvin**
Transition off of Certificate Transparency onto Trillian Logs.

- Trillian does not support Signed Certificate Timestamps.
- Removes CT appender interface.
- Configures servers to explicitly specify the logID for the backend.

The next PR will update the client and server setup scripts

### CODE REVIEW

File: core/client/kt/verify.go [hunk 1]
```diff
@@ -111,15 +109,6 @@ func (v *Verifier) VerifyGetEntryResponse(ctx context.Context, userID string, in
...
-	sct := new(ct.SignedCertificateTimestamp)
-	if _, err := tls.Unmarshal(in.SmhSct, sct); err != nil {
-		return err
-	}
-	if err := v.log.VerifySCT(ctx, in.GetSmh(), sct); err != nil {
-		Vlog.Printf("✗ Signed Map Head CT inclusion proof verification failed.")
-		return err
-	}
-	Vlog.Printf("✓ Signed Map Head CT inclusion proof verified.")
+	// TODO: Verify inclusion proof.
```
Comment (cesarghali) [inline core/client/kt/verify.go]: This reduces the security of the system, do you know when we're planning on adding this verification step?

File: core/keyserver/keyserver.go [hunk 2]
```diff
@@ -77,9 +79,9 @@ func New(committer commitments.Committer, queue queue.Queuer, tree tree.Sparse,
 func (s *Server) GetEntry(ctx context.Context, in *tpb.GetEntryRequest) (*tpb.GetEntryResponse, error) {
 
 	var smh ctmap.SignedMapHead
-	epoch, _, err := s.appender.Latest(ctx, &smh)
+	epoch, err := s.sths.Latest(ctx, s.logID, &smh)
 	if err != nil {
-		log.Printf("Cannot get SMH: %v", err)
+		log.Printf("sth.Latest(%v): %v", s.logID, err)
```
Comment (cesarghali) [inline core/keyserver/keyserver.go]: s/sth/sths

File: cmd/keytransparency-server/frontend.go [hunk 3]
```diff
@@ -193,15 +199,21 @@ func main() {
...
 		log.Fatalf("Failed to create SQL history: %v", err)
 	}
-	sths, err := appender.New(context.Background(), sqldb, *mapID, *mapLogURL, nil)
+
+	tlog, err := config.LogClient(*logID, *logURL, *logPubKey)
 	if err != nil {
-		log.Fatalf("Failed to create appender: %v", err)
+		log.Fatalf("LogClient(%v, %v, %v): %v", *logID, *logURL, *logPubKey, err)
+	}
+	admin := admin.NewStatic()
```
Comment (cesarghali) [inline cmd/keytransparency-server/frontend.go]: The name of this variable is a bit confusing, how about `static` instead, which is what `NewStatic` returns?

File: cmd/keytransparency-signer/backend.go [hunk 4]
```diff
@@ -99,10 +105,15 @@ func main() {
 		log.Fatalf("Failed to create SQL history: %v", err)
 	}
 	mutator := entry.New()
-	sths, err := appender.New(context.Background(), sqldb, *mapID, *mapLogURL, nil)
+	tlog, err := config.LogClient(*logID, *logURL, *logPubKey)
 	if err != nil {
-		log.Fatalf("Failed to create STH appender: %v", err)
+		log.Fatalf("LogClient(%v, %v, %v): %v", *logID, *logURL, *logPubKey, err)
+	}
+	admin := admin.NewStatic()
```
Comment (cesarghali) [inline cmd/keytransparency-signer/backend.go]: Same here.

File: core/keyserver/keyserver_test.go [hunk 5]
```diff
@@ -116,7 +121,11 @@ func addProfiles(count int, c *fakeCommitter, st *fakeSparseHist, a *fakeAppende
 		committed := &tpb.Committed{Data: pData}
 		c.M[string(commitment)] = committed
 		st.M[int64(i)] = commitment
-		a.CurrentEpoch = int64(i)
+
+		smh := new(ctmap.SignedMapHead)
+		if err := sths.Write(context.Background(), 0, int64(i), smh); err != nil {
```
Comment (cesarghali) [inline core/keyserver/keyserver_test.go]: Please create the context in the test function and pass it down to this one.

File: core/keyserver/keyserver_test.go [hunk 6]
```diff
@@ -56,9 +62,14 @@ func TestListEntryHistory(t *testing.T) {
...
 		st := &fakeSparseHist{make(map[int64][]byte)}
-		a := &fakeAppender{0, 0}
-		srv := New(c, fakeQueue{}, st, a, fakePrivateKey{}, fakeMutator{}, authentication.NewFake(), fakeFactory{})
-		if err := addProfiles(profileCount, c, st, a); err != nil {
+		fakeLog := fake.NewFakeTrillianClient()
+		admin := admin.NewStatic()
+		if err := admin.AddLog(logID, fakeLog); err != nil {
+			t.Fatalf("failed to add log to admin: %v", err)
+		}
+		sths := appender.NewTrillian(admin)
```
Comment (cesarghali) [inline core/keyserver/keyserver_test.go]: Can you please mock `sths` similar to what we did with `fakeAppender`? I don't think testing key server depends on having a real `sths` object. Therefore, a fake `sths` does not need to do anything other than implementing the interface.

Same thing if this applies to other tests in this file.

File: core/client/kt/verify.go (hunk 1 above)
Comment (gdbelvin) [inline core/client/kt/verify.go]: I'll add this in the next PR.  To add it here would require expanding this PR to also include all the changes to the client.  For ease of review, I've broken this up.

File: core/keyserver/keyserver.go (hunk 2 above)
Comment (gdbelvin) [inline core/keyserver/keyserver.go]: done

File: cmd/keytransparency-server/frontend.go (hunk 3 above)
Comment (gdbelvin) [inline cmd/keytransparency-server/frontend.go]: done

File: cmd/keytransparency-signer/backend.go (hunk 4 above)
Comment (gdbelvin) [inline cmd/keytransparency-signer/backend.go]: done

File: core/keyserver/keyserver_test.go (hunk 6 above)
Comment (gdbelvin) [inline core/keyserver/keyserver_test.go]: Let's chat about this on GVC. 

This test is using a mock - `fakeLog`. Perhaps there's a better way to setup multi-tennant support - or delay multi tenant support till later.

File: core/keyserver/keyserver_test.go (hunk 5 above)
Comment (gdbelvin) [inline core/keyserver/keyserver_test.go]: done

**Comment by gdbelvin** [comment]
Rebased with comments addressed, PTAL

File: core/keyserver/keyserver_test.go (hunk 6 above)
Comment (cesarghali) [inline core/keyserver/keyserver_test.go]: That's right. However, the `fakeLog` actually has some functionality which is not needed to test this package, but is needed for integration test. I was thinking the mocks will immediately return `nil`. We can talk about this more, though.

**Pull request merged**